woocommerce/woocommerce-android
# [Tooling] Update to bundler 2.2.10 to get security fix and match CI version

## Why?

CircleCI seem to recently have updated to `bundler` version `2.2.10` (we were using `2.1.4` before) on their CI images, probably due to the [important security issue that got fixed in 2.2.10](https://github.com/rubygems/rubygems/blob/master/bundler/CHANGELOG.md#security-fixes).

That version of bundler generate a `Gemfile.lock` that is slightly different from the one we had (especially to separate gems per source, one of the changes that avoids the security exploit) and also seems to resolve the gems slightly differently, leading to running 2.2.10's `bundle install` on the `Gemfile.lock` that was generated using bundler 2.1.4 to fail to resolve the dependency for `rake`

```
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundler could not find compatible versions for gem "rake":
  In snapshot (Gemfile.lock):
    rake (= 12.3.3)
  In Gemfile:
    fastlane (~> 2) was resolved to 2.171.0, which depends on
google-cloud-storage (>= 1.15.0, < 2.0.0) was resolved to 1.29.2, which
depends on
        digest-crc (~> 0.4) was resolved to 0.6.3, which depends on
          rake (>= 12.0.0, < 14.0.0)
    fastlane-plugin-wpmreleasetoolkit was resolved to 0.14.0, which depends on
      rake (~> 12.3)
    fastlane-plugin-wpmreleasetoolkit was resolved to 0.14.0, which depends on
      rake-compiler (~> 1.0) was resolved to 1.1.1, which depends on
        rake
```

## History of Debugging Steps

 - I did a `gem install bundler` on my Mac to update `bundler` to `2.2.10`
 - Then I ran `bundle install`, and confirmed the dependency resolution issue we had on CI with the same 2.2.10
 - Then ran `bundle update` to update the `Gemfile.lock` snapshot, which fixed the issue
 - But `rm -rf vendor/bundle` then `bundle install` re-triggered it, because our release-toolkit Plugin was outside the source block and thus not using the rubygems source, trying to pull its transitive dependencies from local and cache only instead

After reading [the whole thread in bundler's PR](https://github.com/rubygems/rubygems/pull/3655/files) I finally found out that the way we configured our `source` in our Gemfile affected the resolution of transitive dependencies of the plugin and that we should instead only use a global source from now on.

## How?

 - Updated the bundler cache keys on CI to ensure the cache was invalidated (in the end it was not needed, as later in my investigation I ended up updating the `Gemfile.lock` which invalidated the cache anyway, but didn't know that in my earlier attempts)
 - Disabled the use of source blocks and only provided the single `source 'rubygems.org'` as a single global source, [which now allows the plugin to also pull the transitive dependencies from the single global source declaration](https://github.com/rubygems/rubygems/pull/3655/files#diff-48539cd5ac1a26847a9872ff68948c08648586940cba3e3fa9725f1de77e083aR24).

## To Test

 - Wait for CI to go green on the first build (will rebuild the cache due to new `Gemfile.lock` checksum)
 - Ensure another build after the first one (which will use the cache) is still green (that's when the failure started to appeared before that fix)
 - Test to apply the same commits and fix on other failing PRs (currently trying on https://github.com/woocommerce/woocommerce-android/pull/3586)

## Deploying the fix everywhere

 - I already did the same fix on WPAndroid [in this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14082)
 - I plan to try to update the WCiOS and WPiOS repos with 2.2.10 too
 - Then I'll P2 about it
